nameserver/nameserver.go

```diff
@@ -250,41 +261,32 @@ func handleRequest(ctx context.Context, w dns.ResponseWriter, query *dns.Msg) er
 	// TODO: FilterDNSResponse also sets a connection verdict
 	if rrCache == nil {
 		tracer.Infof("nameserver: %s implicitly denied by filtering the dns response, returning nxdomain", conn)
-		returnNXDomain(w, query)
+		returnNXDomain(w, query, conn.Reason)
 		conn.Block("DNS response filtered")
 		return nil
 	}
 
-	// save IP addresses to IPInfo
-	for _, rr := range append(rrCache.Answer, rrCache.Extra...) {
-		switch v := rr.(type) {
-		case *dns.A:
-			ipInfo, err := resolver.GetIPInfo(v.A.String())
-			if err != nil {
-				ipInfo = &resolver.IPInfo{
-					IP:      v.A.String(),
-					Domains: []string{q.FQDN},
-				}
-				_ = ipInfo.Save()
-			} else {
-				added := ipInfo.AddDomain(q.FQDN)
-				if added {
-					_ = ipInfo.Save()
-				}
-			}
-		case *dns.AAAA:
-			ipInfo, err := resolver.GetIPInfo(v.AAAA.String())
-			if err != nil {
-				ipInfo = &resolver.IPInfo{
-					IP:      v.AAAA.String(),
-					Domains: []string{q.FQDN},
-				}
-				_ = ipInfo.Save()
-			} else {
-				added := ipInfo.AddDomain(q.FQDN)
-				if added {
-					_ = ipInfo.Save()
-				}
+	updateIPsAndCNAMEs(q, rrCache, conn)
+
+	// if we have CNAMEs and the profile is configured to filter them
+	// we need to re-check the lists and endpoints here
+	if conn.Process().Profile().FilterCNAMEs() {
+		conn.Entity.ResetLists()
+		conn.Entity.EnableCNAMECheck(true)
+
+		result, reason := conn.Process().Profile().MatchEndpoint(conn.Entity)
+		if result == endpoints.Denied {
+			conn.Block("endpoint in blocklist: " + reason)
+			returnNXDomain(w, query, conn.Reason)
+			return nil
+		}
+
+		if result == endpoints.NoMatch {
+			result, reason = conn.Process().Profile().MatchFilterLists(conn.Entity)
+			if result == endpoints.Denied {
+				conn.Block("endpoint in filterlists: " + reason)
+				returnNXDomain(w, query, conn.Reason)
+				return nil
```

Comment:
1) Shouldn't `updateIPsAndCNAMEs` come after the CNAME filtering? Because if we block the request, the process in question should not try to access these domains via the IPs.
2) Which brings me to the question if it would make sense to move the CNAME filtering to `firewall.FilterDNSResponse` where the other filtering logic is. Would be nice to have everything in one place. (I know this was a difficult and time constrained work here)

Reply:
> 1. Shouldn't updateIPsAndCNAMEs come after the CNAME filtering? Because if we block the request, the process in question should not try to access these domains via the IPs.

updateIPsAndCNAMEs aggregates the CNAMEs that were used to resolve the requested domain to the IP and updates the connections Entity as well as the `IPInfo` (or creates it). That aggregation and update must happen before the CNAME filtering. We coulde move the `IPInfo` update after the CNAME filtering but I'm not sure if there's a benefit. If we block the request the process wouldn't receive the answer and thus don't know the IP to connect to. Or did I get something wrong here?

> 2. Which brings me to the question if it would make sense to move the CNAME filtering to firewall.FilterDNSResponse where the other filtering logic is. Would be nice to have everything in one place. (I know this was a difficult and time constrained work here)

I'm not sure, we don't actually "filter out" CNAME resource records form the DNS anwser section like `FilterDNSResponse` does. We just check if one of the CNAMEs is forbidden and deny the whole request here. `FilterDNSResponse` removes dedicated RR entries from the response and only causes the response to be dropped/blocked if not IPs are left over. 
For simplicity we could/should move that rechecking into a dedicated function but IMHO it doesn't fit into `FilterDNSResponse`.